### DISCUSSION TITLE
Error when update crazyradio-firmware

### DISCUSSION THREAD

**Discussion by nfu64967512** [bitcraze/discussions#1104] [Q&A]
Hello,

I am currently working on a Crazyswarm drone performance project and have successfully completed the initial installation and testing phases. I've configured the channels and addresses for the drones, for example, 0/80/2M:0xE7E7E7E701, 0/90/2M:0xE7E7E7E702, and so on. However, I've run into some issues while attempting to update the Crazyradio 2.0 firmware.

I'm using BitcrazeVM on Xubuntu 20.04, and here are the steps I followed, based on my notes:

sudo apt-get install sdcc binutils
git clone https://github.com/bitcraze/crazyradio-firmware.git
cd ~/crazyradio-firmware/firmware
make CRPA=1 then cd ..
python3 usbtools/launchBootloader.py (This is where I started encountering errors)
sudo python3 ~/crazyradio-firmware/usbtools/nrfbootload.py flash bin/cradio.bin
source ~/catkin_ws/devel/setup.bash
rosrun crazyflie_tools scan -v
During the actual execution, I faced several problems:

When running python3 usbtools/launchBootloader.py, I received the following error:

Launch bootloader Traceback (most recent call last):
  File "usbtools/launchBootloader.py", line 95, in <module>
    sys.exit(launchBootloader(True))
  File "usbtools/launchBootloader.py", line 61, in launchBootloader
    handle.controlMsg(0x40, 0xFF, (), value=0, index=0, timeout=5000)
  File "/home/bitcraze/.local/lib/python3.8/site-packages/usb/legacy.py", line 207, in controlMsg
    return self.dev.ctrl_transfer(
  File "/home/bitcraze/.local/lib/python3.8/site-packages/usb/core.py", line 1082, in ctrl_transfer
    ret = self._ctx.backend.ctrl_transfer(
  File "/home/bitcraze/.local/lib/python3.8/site-packages/usb/backend/libusb1.py", line 893, in ctrl_transfer
    ret = _check(self.lib.libusb_control_transfer(
  File "/home/bitcraze/.local/lib/python3.8/site-packages/usb/backend/libusb1.py", line 604, in _check
    raise USBError(_strerror(ret), ret, _libusb_errno[ret])
usb.core.USBError: [Errno 32] Pipe error

On a subsequent attempt, I got:

Error!, cannot find the CrazyRadio USB dongle!
After these errors, my Crazyradio 2.0 started blinking slowly in red (which I believe indicates it's in bootloader mode), but the behavior seems abnormal. Moreover, the Crazyradio 2.0 disconnects from the VM, and I have to manually re-select it in the VM's USB devices list. After reselection, it shows a drive-like interface, prompting me to drag a .uf2 file into it. However, dragging the file results in a brief white light and then disconnection.

Ignoring these errors and proceeding to the next step(bitcraze@ubuntu:~/crazyradio-firmware$ sudo python3 ~/crazyradio-firmware/usbtools/nrfbootload.py flash bin/cradio.bin
) results in:

Error: cannot found the nRF24LU1 bootloader!

Additionally, when checking the version without the Crazyflie turned on:

Found Crazyradio with version 3.2
No Crazyflie found. Did you specify the correct address?
No Crazyflie connected via USB cable found.

I am seeking advice on how to resolve these issues. Thank you for taking the time to look into my problem.

**Comment by knmcguire** [comment]
Hi! Sorry for the late reply!

It seems that you actually tried to flash the crazyradio 1 firmware on there and not crazyradio 2 firmware. So that is why you get the 'can't find the crazyradio' error.

Please look at this tutorial: https://www.bitcraze.io/documentation/tutorials/getting-started-with-crazyradio-2-0/

**Reply by knmcguire** [reply]
I also see that this is what is given in the Crazyswarm2 documentation as well: https://imrclab.github.io/crazyswarm2/installation.html#updating

**Reply by nfu64967512** [reply]
Hi! I have installed "crazyradio2-CRPA-emulation-1.1.uf2" and drag it to the crazyradio's flie. The light is become white form red so I think it's done the update. But every time I try to update crazyradio's firmware is not success. When I plug the crazyradio2.0 in and start the bootloader by bootloader button and it shows the crazyradio's file window like portable device, so I do the same thing again: pull the uf2 file to the crazyraio2.0 and finish the update. But when I start the bootloader again , it shows up the same window again! And I can't start flash the firmware still just like I mentioned above.
BTW  maybe  I don't have to do the crazyradio-firmware update like :https://github.com/bitcraze/crazyradio-firmware"?
Thanks for the help!

**Reply by nfu64967512** [reply]
I have considered that use crazyswarm2 instead, but my seniors are not  recommended because of ROS2 is not ready for beginner to use , so I still use crazyswarm1 by LPS position still.
Thanks for help!

**Reply by knmcguire** [reply]
Are you sure that it doesn't work? Putting the crazyradio in bootloader mode is suppose to show this window that you are describing. And if the led flashes to white it should have worked.

If you try to connect to the crazyflie with the cfclient, does that work perhaps? https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/

This way we can eliminate any issues related to ROS

**Reply by knmcguire** [reply]
btw, ROS2 is pretty mature now and if you are a beginner, I would recommend you to already start using that, since ROS 1 support will disappear in 2 years. Moreover, support for crazyswarm1 is very minimal now and the development is now mainly focused on crazyswarm2 (this is their discussion forum where you can also get help: https://github.com/IMRCLab/crazyswarm2/discussions)

**Reply by nfu64967512** [reply]
OK now I will do crazyswarm2! if there is any problem I will post on the https://github.com/IMRCLab/crazyswarm2/discussions 
I will close this discussion.

**Reply by knmcguire** [reply]
great! Also from both the ROS community and crazyswarm2, there is more support available as well for ROS2 so in that sense more 'beginner friendly' as well :)

**Comment by nfu64967512** [comment]
Yes! after pull the .uf2 file into crazyraido I can use my crazyradio2.0 on cfclient and config channel and radio address also connect cf successfully. 
Let me figure it out, so when I use "rosrun crazyflie_tools scan -v" I should see "Found Crazyradio with version 99.55" instead of my result "Found Crazyradio with version 3.2"?  
Sorry for lots of problem , still thanks for ur help!

**Reply by knmcguire** [reply]
If the cfclient works then the crazyradio 2 is functional! 

I believe with Crazyradio 2, you are supposed to see `Found Crazyradio with version 3.2`. It might be that crazyswarm (1) is not compatible with crazyradio 2 due to this check. @whoenig, is this correct what I'm saying?

**Reply by whoenig** [reply]
It's possible that Crazyswarm1 shows an error message. Removing the check should work, though. (Of course, the recommended action is to switch to Crazyswarm2, which officially supports Crazyradio 2).

**Reply by nfu64967512** [reply]
Thanks to all of you! I'm in the progress of crazyswarm2! And do this note to update crazyradio2.0 :
1. git clone https://github.com/bitcraze/crazyradio2-firmware.git
2. sudo apt update
    sudo apt install python3-pip
3. pip3 install adafruit-nrfutil
4. put the button and insert the crazyraio2.0, and pull the newest.uf2 file to instead of current.uf2 , and I see the light is turning white from red.
5. adafruit-nrfutil dfu genpkg --dev-type 0x0052 --application build/zephyr/crazyradio2.hex crazyradio2.zip
adafruit-nrfutil dfu serial --package crazyradio2.zip -p /dev/ttyACM0 -b 115200
## I am confuse the step5.'s path, it's always say I'm wrong with path. 
I think above is done the crazyradio2.0-firmware update.
Thanks for help!

**Reply by knmcguire** [reply]
I don't think you should use those instructions in your case. That is only if you want to install the latest development version so it is for development only. You should follow [the tutorial instructions](https://www.bitcraze.io/documentation/tutorials/getting-started-with-crazyradio-2-0/).
